packages/jupyter-ai/jupyter_ai/personas/base_persona.py

```diff
@@ -0,0 +1,189 @@
+        log: Logger,
+    ):
+        self.ychat = ychat
+        self.manager = manager
+        self.config = config
+        self.log = log
+        self.awareness = PersonaAwareness(
+            ychat=self.ychat, log=self.log, user=self.as_user()
+        )
+
+        self.ychat.set_user(self.as_user())
+
+    ################################################
+    # abstract methods, required by subclasses.
+    ################################################
+    @property
+    @abstractmethod
+    def defaults(self) -> PersonaDefaults:
+        """
+        Must return a PersonaDefaults object that represents the default
+        settings of this persona.
+        """
+        pass
+
+    @abstractmethod
+    async def process_message(self, message: Message) -> None:
+        """
+        Processes a new message. Access history and write/stream messages via
+        `self.ychat`.
+        """
+        # support streaming
+        # handle multiple processed messages concurrently (if model service allows it)
+        pass
+
+    ################################################
+    # base class methods, available to subclasses.
+    ################################################
+    @property
+    def id(self) -> str:
+        """
+        Return a unique ID for this persona, which sets its username in the
+        `User` object shared with other collaborative extensions.
+
+        - This ID is guaranteed to be identical throughout this object's
+        lifecycle.
+
+        - The ID is guaranteed to follow the format
+        `jupyter-ai-personas::<package-name>::<persona-class-name>`.
+            - The 'jupyter-ai-personas' prefix allows consumers to easily
+            distinguish AI personas from human users.
+
+            - For example, 'Jupyternaut' always has the ID
+            `jupyter-ai-personas::jupyter-ai::JupyternautPersona`.
+
+        - The ID must be unique, so if a package provides multiple personas,
+        their class names must be unique. Renaming the persona class changes the
+        ID of that persona, so you should also avoid renaming it if possible.
+        """
+        package_name = self.__module__.split(".")[0]
+        class_name = self.__class__.__name__
+        return f"jupyter-ai-personas::{package_name}::{class_name}"
+
+    @property
+    def name(self) -> str:
+        return self.defaults.name
+
+    @property
+    def avatar_path(self) -> str:
+        return self.defaults.avatar_path
+
+    @property
+    def system_prompt(self) -> str:
+        return self.defaults.system_prompt
+
+    def as_user(self) -> User:
+        return User(
+            username=self.id,
+            name=self.name,
+            display_name=self.name,
+            avatar_url=self.avatar_path,
+        )
+
+    def as_user_dict(self) -> Dict[str, Any]:
+        user = self.as_user()
+        return asdict(user)
+
+    async def forward_reply_stream(self, reply_stream: "AsyncIterator"):
```

Comment:
question: do we use typings for async methods in the code base?

It would be something like `Awaitable[None]`

Reply:
Done in latest commit! Thanks for the suggestion.

Reply:
Actually the return type is just `None`. `mypy` automatically wraps return values in `Coroutine[Any, Any, *]`. This is what happens when I annotate the return type as just `None`:

<img width="777" alt="Screenshot 2025-05-19 at 3 30 47 PM" src="https://github.com/user-attachments/assets/31cfaf98-c815-4c27-9a2e-0f75842fb5ec" />